code/iaas/cluster-logic/src/main/java/io/cattle/iaas/cluster/service/impl/ClusterManagerImpl.java

```diff
@@ -0,0 +1,210 @@
+
+            Map<String, Object> params = new HashMap<>();
+            params.put(InstanceConstants.FIELD_NETWORK_IDS, Lists.newArrayList(getNetworkIds(managingHost)));
+            params.put(InstanceConstants.FIELD_REQUESTED_HOST_ID, managingHost.getId());
+            params.put(InstanceConstants.FIELD_PORTS, Lists.newArrayList(clusterServerPort + ":" + clusterServerPort + "/tcp"));
+
+            clusterServerInstance = agentInstanceFactory
+                    .newBuilder()
+                    .withAccountId(managingHost.getAccountId())
+                    .withZoneId(managingHost.getZoneId())
+                    .withPrivileged(true)
+                    .withUri(getUri(cluster, managingHost))
+                    .withName(CLUSTER_INSTANCE_NAME.get())
+                    .withImageUuid(getImageUuid(cluster, managingHost))
+                    .withParameters(params)
+                    .withSystemContainerType(SystemContainer.ClusterAgent)
+                    .build();
+        } else {
+            start(clusterServerInstance);
+        }
+
+        return clusterServerInstance;
+    }
+
+    @Override
+    public Host getClusterForInstance(Instance clusterServerInstance) {
+        if (!isClusterServerInstance(clusterServerInstance)) {
+            return null;
+        }
+        Agent agent = objectManager.loadResource(Agent.class, clusterServerInstance.getAgentId());
+
+        // get cluster id from agent uri
+        String uri = agent.getUri();
+        String[] result = uri.split("clusterId|&");
+        Long clusterId = Long.valueOf(result[1].substring(result[1].indexOf("=") + 1));
+        return objectManager.loadResource(Host.class, clusterId);
+    }
+
+    private boolean isClusterServerInstance(Instance instance) {
+        if (instance.getAgentId() == null) {
+            return false;
+        }
+        String type = instance.getSystemContainer();
+        return (type != null && type.equalsIgnoreCase(SystemContainer.ClusterAgent.name()));
+    }
+
+    private String getImageUuid(Host cluster, Host managingHost) {
+        return getImagePrefix(cluster, managingHost) + CLUSTER_IMAGE_NAME.get();
+    }
+
+    private String getUri(Host cluster, Host managingHost) {
+        return String.format("%s?clusterId=%d&managingHostId=%d",
+                getConnectionPrefix(cluster, managingHost) + "///", cluster.getId(), managingHost.getId());
+    }
+
+    private String getUri(Host cluster) {
+        return getUri(cluster, getManagingHost(cluster));
+    }
+
+    private String getConnectionPrefix(Host cluster, Host managingHost) {
+        return objectManager.isKind(managingHost, "sim") ? "sim:" : "delegate:";
+    }
+
+    private String getImagePrefix(Host cluster, Host managingHost) {
+        return objectManager.isKind(managingHost, "sim") ? "sim:" : "docker:";
+    }
+
+    private Long getNetworkIds(Host managingHost) {
```

Comment:
I extracted a method for retrieving networkId and changed its argument to generic object, as a part of the PR https://github.com/rancherio/cattle/pull/286. So we don't have to repeat this logic for every object type in respective managers. If my PR gets merged first, you re-use this logic here. If not, it should be fixed in a separate PR.